post_office/webhook.py

```diff
@@ -0,0 +1,117 @@
+"""
+A collection of utilities to help process webhook events from email services
+"""
+
+import json
+import logging
+from abc import ABCMeta, abstractmethod
+from enum import Enum
+from typing import Any
+
+from django.http import HttpRequest, HttpResponse, HttpResponseNotFound
+from django.views import View
+from django.views.decorators.csrf import csrf_exempt
+
+from .models import Email
+
+
+logger = logging.getLogger(__name__)
+
+
+class Event(Enum):
+    # Deliverability
+    ACCEPTED = 'accepted'
+    DELIVERED = 'delivered'
+    DEFERRED = 'deferred'
+    HARD_BOUNCE = 'hard_bounce'
+    SOFT_BOUNCE = 'soft_bounced'
+    REJECTED = 'rejected'
+
+    # Engagement
+    OPEN = 'opened'
+    CLICK = 'clicked'
+
+    # Complaints & unsubscribes
+    SPAM_COMPLAINT = 'spam_complaint'
+    UNSUBSCRIBED = 'unsubscribed'
+    RESUBSCRIBED = 'resubscribed'
+
+    # Account
+    ACCOUNT_SUSPENDED = 'account_suspended'
+
+
+class BaseWebhookHandler(View):
+    __metaclass__ = ABCMeta
+
+    @csrf_exempt
+    def dispatch(self, request, *args, **kwargs) -> HttpResponse:
+        # The same method as View.dispatch, but we verify the webhook once
+        # we verify the HTTP method
+        if request.method.lower() in self.http_method_names:
+            if not self.verify_webhook(request, *args, **kwargs):
+                return HttpResponseNotFound()
+            handler = getattr(
+                self,
+                request.method.lower(),
+                self.http_method_not_allowed,
+            )
+        else:
+            handler = self.http_method_not_allowed
+        return handler(request, *args, **kwargs)
+
+    @abstractmethod
+    def verify_webhook(self, request: HttpRequest, *args, **kwargs) -> bool:
+        pass
+
+    def process_event(self, request: HttpRequest, event: Event, email: Email | None = None, *args, **kwargs) -> None:
```

Comment:
Let's keep this simple and have a single `handle_event` to start with. 

```
def handle_event(self, request, event: Event, payload: dict, *args, **kwargs):
```

`event` is whatever event that's sent and can be parsed (including `UNKNOWN`). In the future, we may also pre-parsed metadata such as `subject`, `to` etc, but would like to keep the scope simple for now.

Reply:
NB: Untested since refactor.

Reply:
Oh, and I had to raise the minimum required Python version to 3.10 due to the `argument: Type | None = None` syntax, and only 3.10 and greater supports that.